[VPlan] Expand VPBlendRecipes to select instructions. NFC

lukel97 opened the pull request:
When looking at some EVL tail folded code in SPEC CPU 2017 I noticed we sometimes have both VPBlendRecipes and select VPInstructions in the same plan:

    EMIT vp<%active.lane.mask> = active lane mask vp<%5>, vp<%3>
    EMIT vp<%7> = icmp ...
    EMIT vp<%8> = logical-and vp<%active.lane.mask>, vp<%7>
    BLEND ir<%8> = ir<%n.015> ir<%foo>/vp<%8>
    EMIT vp<%9> = select vp<%active.lane.mask>, ir<%8>, ir<%n.015>

Since a blend will ultimately generate a chain of selects, we could fold the blend into the select:

    EMIT vp<%active.lane.mask> = active lane mask vp<%5>, vp<%3>
    EMIT vp<%7> = icmp ...
    EMIT vp<%8> = logical-and vp<%active.lane.mask>, vp<%7>
    EMIT ir<%8> = select vp<%8>, ir<%foo>, ir<%n.015>

So as a first step, this patch expands blends to a series of select instructions, which may allow them to be simplified further with other select instructions.

lukel97 committed:
[VPlan] Simplify VPBlendRecipes to select instructions

When looking at some EVL tail folded code in SPEC CPU 2017 I noticed we sometimes have both VPBlendRecipes and select VPInstructions in the same plan:

    EMIT vp<%active.lane.mask> = active lane mask vp<%5>, vp<%3>
    EMIT vp<%7> = icmp ...
    EMIT vp<%8> = logical-and vp<%active.lane.mask>, vp<%7>
    BLEND ir<%8> = ir<%n.015> ir<%foo>/vp<%8>
    EMIT vp<%9> = select vp<%active.lane.mask>, ir<%8>, ir<%n.015>

Since a blend will ultimately generate a chain of selects, we could fold the blend into the select:

    EMIT vp<%active.lane.mask> = active lane mask vp<%5>, vp<%3>
    EMIT vp<%7> = icmp ...
    EMIT vp<%8> = logical-and vp<%active.lane.mask>, vp<%7>
    EMIT ir<%8> = select vp<%8>, ir<%foo>, ir<%n.015>

So this patch canonicalizes blends to a series of select instructions, which allows them to be simplified further with other select instructions.

The `BLEND %a, %b, Not(%mask) -> BLEND %b, %a, %mask` optimisation has also been converted to operate on selects.

lukel97 committed:
Update VPlan debug output tests

artagnon reviewed (commented):
Not sure what to think of this, as I don't see a test case with a compelling diff. Can you add one to motivate the patch?

lukel97 commented:
> Not sure what to think of this, as I don't see a test case with a compelling diff. Can you add one to motivate the patch?

In https://github.com/llvm/llvm-project/pull/133993/files#diff-b5b1d6f9b464b1caab9b230ba0cbcc01378b45452e68ad623f530bc64062a12c we're able to take fold away a select from an existing simplification, and in https://github.com/llvm/llvm-project/pull/133993/files#diff-563df92f6b55e48060919bceeefa8d520d46f96cf360b5f1bb62e90c252ab403 we are able to remove a couple of redundant nots. 

But the primary motivator for this is that with RISC-V EVL tail folding, there's a good few places in SPEC CPU 2017 where we have redundant masking. The issue is that we have VPBlendRecipes that have the header mask added into them but we aren't replacing them with vp.selects.

The plan is that by canonicalising VPBlendRecipes to select VPInstructions, we can reuse the EVL simplifcation we already have for select VPInstructions here:

https://github.com/llvm/llvm-project/blob/e873ce31ae0e875081c8e5480c9c4925c97469ce/llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp#L2175-L2187

lukel97 committed:
Mark in planContainsAdditionalSimplifications

This fixes a crash on SPEC CPU 2017 with mismatching legacy + vplan cost models.

I'm really struggling to create a test case for this due to the issue at https://github.com/llvm/llvm-project/issues/147038

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/simplifiy-blend-select

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -928,6 +928,19 @@ InstructionCost VPInstruction::computeCost(ElementCount VF,
                                       Ctx.CostKind);
   }
+  case Instruction::Select: {
+    if (!getUnderlyingValue())
```
```suggestion
    if (!isLiveIn() || !getUnderlyingValue())
```

lukel97 replied:
Is it possible for a VPInstruction to be a live in? A live in has to not have a recipe:

```c++
  /// Returns true if this VPValue is a live-in, i.e. defined outside the VPlan.
  bool isLiveIn() const { return !hasDefiningRecipe(); }
```

artagnon commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7047,6 +7047,13 @@ static bool planContainsAdditionalSimplifications(VPlan &Plan,
+      using namespace VPlanPatternMatch;
+      if (match(&R, m_VPInstruction<Instruction::Select>(
+                        m_VPValue(), m_VPValue(), m_VPValue())) &&
+          isa_and_nonnull<PHINode>(R.getVPSingleValue()->getUnderlyingValue()))
```
I think the same isLiveIn() check is necessary?

lukel97 replied:
I pushed 4911b73a3262ba31b9c868c53ce021b4af3e936f to do this but then it turns out getUnderlyingInstr calls cast<> which asserts on null, so I had to revert it in 78c3a086c64c974519847a53ba83a27eb253360d

artagnon commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1082,6 +1082,15 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
   if (match(Def, m_Select(m_VPValue(), m_VPValue(X), m_Deferred(X))))
     return Def->replaceAllUsesWith(X);
 
+  // select !c, x, y -> select c, y, x
+  VPValue *C;
+  if (match(Def, m_Select(m_Not(m_VPValue(C)), m_VPValue(X), m_VPValue(Y)))) {
+    Def->setOperand(0, C);
+    Def->setOperand(1, Y);
+    Def->setOperand(2, X);
+    return;
+  }
+
```
Can this change be landed separately?

artagnon reviewed (commented):
I saw your follow-up patch, and I think this LGTM, although I feel like I'm missing context.

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/simplifiy-blend-select

lukel97 committed:
Use getUnderlyingInstr

lukel97 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7047,6 +7047,13 @@ static bool planContainsAdditionalSimplifications(VPlan &Plan,
 
+      // VPBlendRecipes are converted to selects and may have been simplified.
+      using namespace VPlanPatternMatch;
+      if (match(&R, m_VPInstruction<Instruction::Select>(
+                        m_VPValue(), m_VPValue(), m_VPValue())) &&
```
Good catch, I didn't notice this at the time. But I tried it out and it turns out m_Select also matches VPWidenSelectRecipes and VPReplicateRecipes too which we don't want to skip

lukel97 committed:
Revert "Use getUnderlyingInstr"

This reverts commit 4911b73a3262ba31b9c868c53ce021b4af3e936f.

Turns out we can't use getUnderlyingInstr since it does a cast<> which asserts on null

artagnon reviewed (approved):
Okay, I've convinced myself that this is a good direction, given the follow-up and context. I never really understood why we weren't lowering blends to primitive instructions, which are easier to reason about, in the first place. LGTM, pending on a second reviewer.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7047,6 +7047,13 @@ static bool planContainsAdditionalSimplifications(VPlan &Plan,
 
+      // VPBlendRecipes are converted to selects and may have been simplified.
+      using namespace VPlanPatternMatch;
+      if (match(&R, m_VPInstruction<Instruction::Select>(
```
Why is this needed? This effectively disables cost checking for any plans with blends? If the blend gets simplified away, the regular checks should already catch this?

fhahn replied:
Argh I see, so we create multiple selects with the same underlying instruction and this is causing the issue.

As lowering blends to selects becomes mandatory, it would probably be better to do it in convertToConcreteRecipes, at least to start with, removing the cost-model issue?

lukel97 replied:
We could do it in convertToConcreteRecipes but then we don't get any of the benefits of the simplifications that occur on selects, both in `VPlanTransforms::simplifyRecipes` and in `optimizeMaskToEVL`. 

Although I'm happy to at least start in convertToConcreteRecipes if you think it'll be ok to move the transform to earlier in a follow up patch?

fhahn replied:
Yes that would probably be good as a first step, to start with removing VPBlendRecipe::execute.

lukel97 committed:
Precisely track blend selects to see if they have been simplified, add test case

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/simplifiy-blend-select

lukel97 committed:
Don't simplify, just expand in convertToConcreteRecipes

lukel97 committed:
Remove redundant setUnderlyingValue, adjust comment

fhahn reviewed (approved):
LGTM, thanks

lukel97 committed:
Move VPBlendRecipe::execute into header

lukel97 committed:
Merge branch 'main' of github.com:llvm/llvm-project into loop-vectorize/simplifiy-blend-select
